File: mobile/android/TeleSol/app/src/main/java/com/telesol/companion/data/BackendService.kt
// ════════════════════════════════════════════════════════════════
// TeleSol Android — Backend Communication Service
// ════════════════════════════════════════════════════════════════

package com.telesol.companion.data

import android.content.Context
import android.content.SharedPreferences
import android.os.BatteryManager
import android.os.Build
import android.provider.Settings
import com.google.gson.Gson
import com.telesol.companion.data.models.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class BackendService(private val context: Context) {

    // ── State ────────────────────────────────────────────────
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val _isRegistered = MutableStateFlow(false)
    val isRegistered: StateFlow<Boolean> = _isRegistered

    private val _isStreaming = MutableStateFlow(false)
    val isStreaming: StateFlow<Boolean> = _isStreaming

    private val _lastCRS = MutableStateFlow(0.0)
    val lastCRS: StateFlow<Double> = _lastCRS

    private val _lastLevel = MutableStateFlow("normal")
    val lastLevel: StateFlow<String> = _lastLevel

    private val _espConnected = MutableStateFlow(false)
    val espConnected: StateFlow<Boolean> = _espConnected

    private val _streamCount = MutableStateFlow(0)
    val streamCount: StateFlow<Int> = _streamCount

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    // ── Config ───────────────────────────────────────────────
    private val prefs: SharedPreferences = context.getSharedPreferences("telesol_prefs", Context.MODE_PRIVATE)
    var backendURL: String
        get() = prefs.getString("backend_url", "http://192.168.1.50:8000") ?: "http://192.168.1.50:8000"
        set(value) = prefs.edit().putString("backend_url", value).apply()

    private val deviceId: String by lazy {
        Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID) ?: "unknown"
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .build()

    private val gson = Gson()
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var streamJob: Job? = null

    // ── Registration ─────────────────────────────────────────
    fun register() {
        val registration = MobileRegistration(
            deviceId = deviceId,
            deviceName = "${Build.MANUFACTURER} ${Build.MODEL}",
            osType = "android",
            osVersion = Build.VERSION.RELEASE
        )

        val body = gson.toJson(registration).toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url("$backendURL/api/mobile/register")
            .post(body)
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                _isConnected.value = false
                _errorMessage.value = "Registration failed: ${e.message}"
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    _isRegistered.value = true
                    _isConnected.value = true
                    _errorMessage.value = null
                } else {
                    _errorMessage.value = "Registration failed: HTTP ${response.code}"
                }
                response.close()
            }
        })
    }

    // ── Stream Data ──────────────────────────────────────────
    fun sendStream(
        noiseDb: Double,
        personCount: Int,
        motionDetected: Boolean,
        latitude: Double?,
        longitude: Double?,
        accuracy: Double?
    ) {
        if (!_isRegistered.value) return

        val batteryManager = context.getSystemService(Context.BATTERY_SERVICE) as? BatteryManager
        val batteryLevel = batteryManager?.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)

        val payload = MobileStreamPayload(
            deviceId = deviceId,
            timestamp = System.currentTimeMillis(),
            noiseDb = noiseDb,
            isCrowdNoise = noiseDb > 60.0,
            personCount = personCount,
            motionDetected = motionDetected,
            gpsLatitude = latitude,
            gpsLongitude = longitude,
            gpsAccuracy = accuracy,
            batteryLevel = batteryLevel
        )

        val body = gson.toJson(payload).toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url("$backendURL/api/mobile/stream")
            .post(body)
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                _isConnected.value = false
                _errorMessage.value = e.message
            }

            override fun onResponse(call: Call, response: Response) {
                _isConnected.value = true
                _streamCount.value += 1

                response.body?.string()?.let { responseBody ->
                    try {
                        val resp = gson.fromJson(responseBody, StreamResponse::class.java)
                        _lastCRS.value = resp.fusedCrs
                        _lastLevel.value = resp.fusedLevel
                        _espConnected.value = resp.espConnected
                    } catch (_: Exception) {}
                }
                response.close()
            }
        })
    }

    // ── Auto-Streaming ───────────────────────────────────────
    fun startStreaming(
        noiseProvider: () -> Double,
        personProvider: () -> Int,
        locationProvider: () -> Triple<Double?, Double?, Double?>
    ) {
        _isStreaming.value = true
        streamJob = scope.launch {
            while (isActive) {
                val noise = noiseProvider()
                val persons = personProvider()
                val (lat, lng, acc) = locationProvider()
                sendStream(noise, persons, persons > 0, lat, lng, acc)
                delay(3000)  // Stream every 3 seconds
            }
        }
    }

    fun stopStreaming() {
        streamJob?.cancel()
        streamJob = null
        _isStreaming.value = false
    }

    // ── Disconnect ───────────────────────────────────────────
    fun disconnect() {
        stopStreaming()
        val body = gson.toJson(mapOf("device_id" to deviceId)).toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url("$backendURL/api/mobile/disconnect")
            .post(body)
            .build()
        try { client.newCall(request).execute().close() } catch (_: Exception) {}
    }

    fun cleanup() {
        scope.cancel()
    }
}

File: mobile/ios/TeleSol/Services/BackendService.swift
// ════════════════════════════════════════════════════════════════
// TeleSol iOS — Backend Communication Service
// ════════════════════════════════════════════════════════════════

import Foundation
import Combine
import UIKit

class BackendService: ObservableObject {
    // MARK: - Published State
    @Published var isConnected = false
    @Published var isRegistered = false
    @Published var isStreaming = false
    @Published var lastCRS: Double = 0.0
    @Published var lastLevel: String = "normal"
    @Published var espConnected = false
    @Published var errorMessage: String?
    @Published var streamCount: Int = 0
    @Published var dashboardData: DashboardData?

    // MARK: - Configuration
    @Published var backendURL: String {
        didSet {
            UserDefaults.standard.set(backendURL, forKey: "backend_url")
        }
    }

    // MARK: - Private
    private var deviceId: String
    private var streamTimer: Timer?
    private var wsTask: URLSessionWebSocketTask?
    private let session = URLSession.shared
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Init
    init() {
        self.backendURL = UserDefaults.standard.string(forKey: "backend_url") ?? "http://192.168.1.50:8000"
        self.deviceId = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
    }

    // MARK: - Registration
    func register(deviceName: String, osVersion: String) {
        let registration = MobileRegistration(
            deviceId: deviceId,
            deviceName: deviceName,
            osType: "ios",
            osVersion: osVersion,
            appVersion: "1.0.0"
        )

        guard let url = URL(string: "\(backendURL)/api/mobile/register") else {
            errorMessage = "Invalid backend URL"
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10

        do {
            request.httpBody = try JSONEncoder().encode(registration)
        } catch {
            errorMessage = "Failed to encode registration: \(error)"
            return
        }

        session.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    self?.errorMessage = "Registration failed: \(error.localizedDescription)"
                    self?.isConnected = false
                    return
                }

                guard let httpResponse = response as? HTTPURLResponse,
                      httpResponse.statusCode == 200 else {
                    self?.errorMessage = "Registration failed: bad response"
                    return
                }

                self?.isRegistered = true
                self?.isConnected = true
                self?.errorMessage = nil

                // Connect to WebSocket for live updates
                self?.connectWebSocket()
            }
        }.resume()
    }

    // MARK: - Stream Data to Backend
    func sendStream(noiseDb: Double, personCount: Int, motionDetected: Bool,
                    latitude: Double?, longitude: Double?, accuracy: Double?) {
        guard isRegistered else { return }

        let payload = MobileStreamPayload(
            deviceId: deviceId,
            timestamp: Int64(Date().timeIntervalSince1970 * 1000),
            noiseDb: noiseDb,
            peakAmplitude: 0.0,
            isCrowdNoise: noiseDb > 60.0,
            personCount: personCount,
            motionDetected: motionDetected,
            gpsLatitude: latitude,
            gpsLongitude: longitude,
            gpsAccuracy: accuracy,
            batteryLevel: Int(UIDevice.current.batteryLevel * 100)
        )

        guard let url = URL(string: "\(backendURL)/api/mobile/stream") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 5

        do {
            request.httpBody = try JSONEncoder().encode(payload)
        } catch { return }

        session.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    self?.isConnected = false
                    self?.errorMessage = error.localizedDescription
                    return
                }

                self?.isConnected = true
                self?.streamCount += 1

                if let data = data,
                   let resp = try? JSONDecoder().decode(StreamResponse.self, from: data) {
                    self?.lastCRS = resp.fusedCrs
                    self?.lastLevel = resp.fusedLevel
                    self?.espConnected = resp.espConnected
                }
            }
        }.resume()
    }

    // MARK: - Auto-Streaming
    func startStreaming(noiseProvider: @escaping () -> Double,
                        personProvider: @escaping () -> Int,
                        locationProvider: @escaping () -> (Double?, Double?, Double?)) {
        isStreaming = true
        streamTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { [weak self] _ in
            let noise = noiseProvider()
            let persons = personProvider()
            let (lat, lng, acc) = locationProvider()
            self?.sendStream(
                noiseDb: noise,
                personCount: persons,
                motionDetected: persons > 0,
                latitude: lat,
                longitude: lng,
                accuracy: acc
            )
        }
    }

    func stopStreaming() {
        streamTimer?.invalidate()
        streamTimer = nil
        isStreaming = false
    }

    // MARK: - WebSocket (live dashboard data)
    private func connectWebSocket() {
        guard let url = URL(string: backendURL.replacingOccurrences(of: "http", with: "ws") + "/ws") else { return }

        wsTask = session.webSocketTask(with: url)
        wsTask?.resume()
        receiveWebSocketMessage()
    }

    private func receiveWebSocketMessage() {
        wsTask?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    if let data = text.data(using: .utf8),
                       let json = try? JSONDecoder().decode(WebSocketMessage.self, from: data) {
                        DispatchQueue.main.async {
                            self?.dashboardData = json.data
                        }
                    }
                default:
                    break
                }
                self?.receiveWebSocketMessage()  // Continue listening

            case .failure:
                DispatchQueue.main.async {
                    self?.isConnected = false
                }
                // Reconnect after delay
                DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
                    self?.connectWebSocket()
                }
            }
        }
    }

    // MARK: - Disconnect
    func disconnect() {
        stopStreaming()
        wsTask?.cancel(with: .goingAway, reason: nil)

        guard let url = URL(string: "\(backendURL)/api/mobile/disconnect") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONEncoder().encode(["device_id": deviceId])
        session.dataTask(with: request).resume()
    }
}

// WebSocket message wrapper
struct WebSocketMessage: Codable {
    let type: String
    let data: DashboardData?
}
